pkg/core/blockchain.go

```diff
@@ -410,6 +413,72 @@ func (bc *Blockchain) init() error {
 	return bc.updateExtensibleWhitelist(bHeight)
 }
 
+// JumpToState is an atomic operation that changes Blockchain state to the one
+// specified by the state sync point p. All the data needed for the jump must be
+// collected by the state sync module.
+func (bc *Blockchain) JumpToState(p uint32) error {
```

Comment:
Can it validate `stateSync` module state before doing anything? It's quite dangerous method API-wise, it must ensure that the jump is possible.

Reply:
It can, but are you sure that we need it? The only place where `JumpToState` is called is
```
func (s *Module) checkSyncIsCompleted() {
	if s.syncStage != headersSynced|mptSynced|blocksSynced {
		return
	}
	s.log.Info("state is in sync",
		zap.Uint32("state sync point", s.syncPoint))
	err := s.bc.JumpToState(s.syncPoint)
	if err != nil {
		s.log.Fatal("failed to jump to the latest state sync point", zap.Error(err))
	}
	s.syncStage = inactive
	s.dispose()
}
```

and it guarantees that the module state is proper.

Currently, I've fixed it with the module state check, but that doesn't look good to me. If you're worried about exposing `(*Blockchain).JumpToState`, then I'd better convert it to a callback and pass it as `(s *Module) Init(currChainHeight uint32, jumpCallback func(p uint32) error)`.

Reply:
I've implemented the following:
1. statesync.Module instance is created directly in `GetStateSyncModule` now.
2. Statesync module is not tied to the Blockchain anymore.
3. (*Blockchain).JumpToState is a callback now, we don't expose it. Was not requested, but I have implemented it in a separate commit. We can drop it if we don't like it.

> This BTW could allow checking the network configuration and just refusing to give stateSync if it can't really work.

I think that statesync module should always be returned from `GetStateSyncModule`, because otherwise the requester should be fond of the current state sync stage, and that's what we'd like to avoid by concentrating the whole state-related logic in one place (in the statesync module).